File: 216_combination_sum_III.cpp
class Solution {
public:
    vector<vector<int>> combinationSum3(int k, int n) {
        vector<vector<int> > result;
        vector<int> path;
        dfs(result,path,n,0,0,k,0);
        return result;
    }
    void dfs(vector<vector<int> >& result, vector<int>& path, int target, int sum, int deep, int count, int start) {
        if (deep==count) {
            if (sum==target) result.push_back(path);
            return;}
        for (int i=start+1;i<10;i++) {
            if (sum+i>target) break;
            path.push_back(i);
            dfs(result,path,target, sum+i,deep+1,count,i);
            path.pop_back();
            }
        
        }
};

File: algorithms/python3/216_combination_sum_III.py
class Solution:
    def combinationSum3(self, k: int, n: int) -> List[List[int]]:

        def dfs(start,k,target,path):
            if len(path)==k and target==0:
                res.append(path)
            else:
                for i in range(start,10):
                    if target<i:
                        break
                    else:
                        dfs(i+1,k,target-i,path+[i])

        res=[]
        dfs(1,k,n,[])
        return res

a=Solution()

print(a.combinationSum3(3,9))

